encoding/ssh/filexfer/packets.go

```diff
@@ -0,0 +1,227 @@
+package filexfer
+
+import (
+	"errors"
+	"fmt"
+	"io"
+)
+
+func newPacketFromType(typ PacketType) (Packet, error) {
+	switch typ {
+	case PacketTypeOpen:
+		return new(OpenPacket), nil
+	case PacketTypeClose:
+		return new(ClosePacket), nil
+	case PacketTypeRead:
+		return new(ReadPacket), nil
+	case PacketTypeWrite:
+		return new(WritePacket), nil
+	case PacketTypeLstat:
+		return new(LstatPacket), nil
+	case PacketTypeFstat:
+		return new(FstatPacket), nil
+	case PacketTypeSetstat:
+		return new(SetstatPacket), nil
+	case PacketTypeFsetstat:
+		return new(FsetstatPacket), nil
+	case PacketTypeOpendir:
+		return new(OpendirPacket), nil
+	case PacketTypeReaddir:
+		return new(ReaddirPacket), nil
+	case PacketTypeRemove:
+		return new(RemovePacket), nil
+	case PacketTypeMkdir:
+		return new(MkdirPacket), nil
+	case PacketTypeRmdir:
+		return new(RmdirPacket), nil
+	case PacketTypeRealpath:
+		return new(RealpathPacket), nil
+	case PacketTypeStat:
+		return new(StatPacket), nil
+	case PacketTypeRename:
+		return new(RenamePacket), nil
+	case PacketTypeReadlink:
+		return new(ReadlinkPacket), nil
+	case PacketTypeSymlink:
+		return new(SymlinkPacket), nil
+	case PacketTypeExtended:
+		return new(ExtendedPacket), nil
+	default:
+		return nil, fmt.Errorf("unexpected request packet type: %v", typ)
+	}
+}
+
+// RawPacket implements the general packet format from draft-ietf-secsh-filexfer-02
+//
+// Defined in https://tools.ietf.org/html/draft-ietf-secsh-filexfer-02#section-3
+type RawPacket struct {
+	Type      PacketType
+	RequestID uint32
+
+	Data Buffer
+}
+
+// Reset clears the pointers and reference-semantic variables of RawPacket,
+// making it suitable to be put into a sync.Pool.
+func (p *RawPacket) Reset() {
+	p.Data.Reset()
+}
+
+// MarshalPacket returns p as a two-part binary encoding of p.
+//
+// The internal p.RequestID is overridden by the reqid argument.
+func (p *RawPacket) MarshalPacket(reqid uint32) (header, payload []byte, err error) {
+	b := NewMarshalBuffer(p.Type, reqid, 0)
+
+	return b.Packet(p.Data.Bytes())
+}
+
+// MarshalBinary returns p as the binary encoding of p.
+func (p *RawPacket) MarshalBinary() ([]byte, error) {
+	return ComposePacket(p.MarshalPacket(p.RequestID))
+}
+
+// UnmarshalFrom decodes a RawPacket from the given Buffer into p.
+func (p *RawPacket) UnmarshalFrom(buf *Buffer) error {
+	typ, err := buf.ConsumeUint8()
+	if err != nil {
+		return err
+	}
+
+	p.Type = PacketType(typ)
+
+	if p.RequestID, err = buf.ConsumeUint32(); err != nil {
+		return err
+	}
+
+	p.Data = *buf
```

Comment:
This too takes ownership of the underlying buffer.

Reply:
Like the other function, I’ll document this.